ui/src/shared/components/ThresholdsList.js

```diff
@@ -0,0 +1,192 @@
+        randomValue = _.round(_.random(minValue, maxValue, true), 2)
+      } while (_.includes(colorsValues, randomValue))
+    }
+
+    const newThreshold = {
+      type: thresholdsListType,
+      id: uuid.v4(),
+      value: randomValue,
+      hex: GAUGE_COLORS[randomColor].hex,
+      name: GAUGE_COLORS[randomColor].name,
+    }
+
+    const updatedColors = _.sortBy(
+      [...thresholdsListColors, newThreshold],
+      color => color.value
+    )
+
+    handleUpdateThresholdsListColors(updatedColors)
+    onResetFocus()
+  }
+
+  handleDeleteThreshold = threshold => () => {
+    const {handleUpdateThresholdsListColors, onResetFocus} = this.props
+    const thresholdsListColors = this.props.thresholdsListColors.filter(
+      color => color.id !== threshold.id
+    )
+    const sortedColors = _.sortBy(thresholdsListColors, color => color.value)
+
+    handleUpdateThresholdsListColors(sortedColors)
+    onResetFocus()
+  }
+
+  handleChooseColor = threshold => chosenColor => {
+    const {handleUpdateThresholdsListColors} = this.props
+
+    const thresholdsListColors = this.props.thresholdsListColors.map(
+      color =>
+        color.id === threshold.id
+          ? {...color, hex: chosenColor.hex, name: chosenColor.name}
+          : color
+    )
+
+    handleUpdateThresholdsListColors(thresholdsListColors)
+  }
+
+  handleUpdateColorValue = (threshold, value) => {
+    const {handleUpdateThresholdsListColors} = this.props
+
+    const thresholdsListColors = this.props.thresholdsListColors.map(
+      color => (color.id === threshold.id ? {...color, value} : color)
+    )
+
+    handleUpdateThresholdsListColors(thresholdsListColors)
+  }
+
+  handleValidateColorValue = (threshold, targetValue) => {
+    const {thresholdsListColors} = this.props
+    const sortedColors = _.sortBy(thresholdsListColors, color => color.value)
+
+    return !sortedColors.some(color => color.value === targetValue)
+  }
+
+  handleSortColors = () => {
+    const {thresholdsListColors, handleUpdateThresholdsListColors} = this.props
+    const sortedColors = _.sortBy(thresholdsListColors, color => color.value)
+
+    handleUpdateThresholdsListColors(sortedColors)
+  }
+
+  render() {
+    const {thresholdsListColors, showListHeading} = this.props
+    const disableAddThreshold = thresholdsListColors.length > MAX_THRESHOLDS
+
+    const thresholdsListClass = `thresholds-list${showListHeading &&
+      ' graph-options-group'}`
+
+    return (
+      <div className={thresholdsListClass}>
+        {showListHeading && <label className="form-label">Thresholds</label>}
+        <button
+          className="btn btn-sm btn-primary"
+          onClick={this.handleAddThreshold}
+          disabled={disableAddThreshold}
+        >
+          <span className="icon plus" /> Add Threshold
+        </button>
+        {thresholdsListColors.map(
+          color =>
+            color.id === THRESHOLD_TYPE_BASE
+              ? <div className="threshold-item" key={color.id}>
```

Comment:
The colors ID is generated via uuid when the color is created

Reply:
Soo. one of the bugs Jared was addressing earlier in the week was about duplicated keys. Keys are something that react needs to generate elements by mapping over data.- but we don't want anything on the page to have duplicate keys. My concern isn't that the number was generated using uuidv4 or any other kind of random number generator somewhere along the line. My concern is that they are guaranteed to be unique. In this case you are not using color.id as a key on two separate things, since the other component that uses color.id is on the other side of a terniary.  BUT what if in a refactor, we decide to show both of these, then they would have the same key since they are depending on the underlying data. I think all keys should be rendered with uuids. This is my opinion.